Changelogs v1.12.0 (#73)
meilisearch/engine-team: meilisearch-release-changelog/v1.12.0.md

Possible values are:

- `"disabled"`: deactivate the prefix search completely
- `"indexingTime"`: activate the processing of the prefixes (default Meilisearch behavior)

Disabling the prefix search implies that, during a search, the word `hello` will not be retrieved anymore with the query `he`. But the prefixes will not be processed during the indexing, saving time.

## More visibility around indexing tasks

In order to give more visibility around indexing processing around your `task`, especially when indexing big documents, new routes have been introduced:
- `GET /batches/:uid`: returns a specific `batch` object
- `GET /batches`: return a list of `batch` objects

The same query parameters than `GET /tasks` route can be used to apply filtering.

TBD: an example of batch object

Also, a new field is introduced in the task object: `batch`.

TBD: an example of the new `task` object, one `progress` in implemented.

Done by @irevoire in #5060, #5070, #5080

## Other improvements

* Introduce the `reverse` query parameter for `GET /tasks` route set to `false` by default. If set to `true`, then the tasks will be returned in reversed order (the oldest first). Done by @irevoire in #5048.

> * Make matches consider phrases as a single `Match` by @flevi29 in #4928

Reviewer: ```suggestion
* Phrase searches with`showMatchesPosition` set to `true` give a single location for the whole phrase @flevi29 in #4928
```

> * Adds new metrics to prometheus by @PedroTurik in #5044

Reviewer: ```suggestion
* New Prometheus metrics by @PedroTurik in #5044
```

> * Add `indices` field to `_matchesPosition` to specify where in an array a match comes from, by @LukasKalbertodt in #5005

Reviewer: ```suggestion
* When a query finds matching terms in document fields with array values,  Meilisearch now includes an `indices` field to `_matchesPosition` specifying which array elements contain the matches by @LukasKalbertodt in #5005
```

> * ⚠️ Breaking for users of the experimental `vectorStore` feature: The field distribution no longer contains `_vectors`. Its value used to be incorrect, and fixing it gives it a very uninteresting value, as all documents in an index with an embedder have a (potentially empty) `_vectors` field. Done as part of #4900

Reviewer: ```suggestion
* ⚠️ Breaking `vectorStore` change: field distribution no longer contains `_vectors`. Its value used to be incorrect, and there is no current use case for the fixed, most likely empty, value. Done as part of #4900
```

# Fixes 🐞

> * Return an appropriate error when primary keys is greater than 512 bytes, by @flevi29 in #4930

Reviewer: ```suggestion
* Return appropriate error when primary key is greater than 512 bytes, by @flevi29 in #4930
```

> * Numbers are now segmented the same way by the tokenizer by @dqkqd in https://github.com/meilisearch/charabia/pull/311

Reviewer: ```suggestion
* Fix issue where numbers were segmented in different ways depending on tokenizer, by @dqkqd in https://github.com/meilisearch/charabia/pull/311
```

* Fix pagination when embedding fails by @dureuill in https://github.com/meilisearch/meilisearch/pull/5063

> * Stop words are not ignored anymore in #5062 by @ManyTheFish

Reviewer: ```suggestion
* Fix issue causing Meilisearch to ignore stop words in some cases by @ManyTheFish in #5062
```

Meilisearch v1.12 introduces significant indexing speed improvements, almost halving the time required to index large datasets. This release also introduces new settings to customize and potentially further increase indexing speed.

<!-- The following lines should NOT be put in the PRE-release changelogs -->
🧰 All official Meilisearch integrations (including SDKs, clients, and other tools) are compatible with this Meilisearch release. Integration deployment happens between 4 to 48 hours after a new version becomes available.

<!-- The following lines should NOT be put in the PRE-release changelogs -->
Some SDKs might not include all new features. Consult the project repository for detailed information. Is a feature you need missing from your chosen SDK? Create an issue letting us know you need it, or, for open-source karma points, open a PR implementing it (we'll love you for that ❤️).

# New features and updates 🔥

## Improve indexing speed

Indexing time for huge dataset import (multiple millions of documents) is divided by two!

> TBD: give more data and metrics about improvements

Reviewer 2: The new default value for the max-indexing-memory parameter is no longer 2/3 of the available memory but 5%. You can find more information on [the related PR](https://github.com/meilisearch/meilisearch/pull/5125).

Reviewer 3: Note, this might change

Author: I will let you update the changelog next week or next-next week

Reviewer 2: No longer true: https://github.com/meilisearch/meilisearch/pull/5141

Use this setting to toggle [facet search](https://www.meilisearch.com/docs/learn/filtering_and_sorting/search_with_facet_filters#searching-facet-values):

```bash
curl \
  -X PUT 'http://localhost:7700/indexes/books/settings/facet-search' \
  -H 'Content-Type: application/json' \
  --data-binary 'true'
```

The default value for `facetSearch` is `true`. When set to `false`, this setting disables facet search for all filterable attributes in an index.

### `prefixSearch`

Use this setting to configure the ability to [search a word by prefix](https://www.meilisearch.com/docs/learn/engine/prefix) on an index:

```bash
curl \
  -X PUT 'http://localhost:7700/indexes/books/settings/prefix-search' \
  -H 'Content-Type: application/json' \
  --data-binary 'disabled'
```

`prefixSearch` accepts one of the following values:

- `"indexingTime"`: enables prefix processing during indexing. This is the default Meilisearch behavior
- `"disabled"`: deactivates prefix search completely

Disabling prefix search means the query `he` will no longer match the word `hello`. This may significantly impact search result relevancy, but speeds up the indexing process.

## New API route: `/batches`

> The new `/batches` endpoint allow you to query information about task batches.

Reviewer 4: ```suggestion

To give more visibility into how tasks are processed and batched together, Meilisearch introduced two new routes that give a view of the batches.
These views will give you some stats on what was processed in the batch and also a detailed view of the progress Meilisearch made on the batch. Finally, an explanation of what happens when a task is marked as processing :tada: 
```

Reviewer: In general I think we should aim to be as brief as possible in the changelog, so users can skim it and quickly assess if anything is important to them. 

Contextualising changes is more the work of the documentation. And contextualising changes in a fun way is more of the work of the blog, because documentarians hate fun.

`GET` `/batches` returns a list of batch objects:

```sh
curl  -X GET 'http://localhost:7700/batches'
```

> 
> This endpoint accepts the same parameters as `GET` `/tasks` route, allowing you to narrow down which batches you want to see.
> 

Reviewer 4: ```suggestion

This endpoint accepts the same parameters and returns the same error codes as the `GET` `/tasks` route, allowing you to narrow down which batches you want to see.
But don't forget, the filters are applied **to the tasks**, not the batches. Which means, querying `GET /batches?status=failed` won't return the batches that failed but will instead returns the batches **containing at least** one failed task.
The same is true for all query parameters. (`uids` specify the `uids` of the tasks, `before_started_at`, etc)
And your API key also impacts the results. A batch won't be returned if it only contains tasks your API key don't let you see.
```

Author: @irevoire is the suggestion of @guimachiavelli correct to you?

Reviewer 4: Hey,

> For some parameters like status=failed I guess it's irrelevant, since AFAIK the whole batch fails if even a single task fails.

No! That's not the case, a single batch can contain both failed and succeeded tasks.
And also, a batch doesn't have a status by itself at all.
This means that you can match the same batch by using the filter `?statuses=succeeded` and `?statuses=failed`.

> I would definitely expect GET /batches?uids=0,2 to return batches with a uid of either 1 or 2, and even more so if GET /tasks?uids=0,2 returns tasks 1 and 2.

Yeah, I too think that's a bit confusing.
The purpose of this naming was that you can copy paste your filter on the `/tasks` route and it just works. There is no « oh I must change the `uid` in `batchUid` », or the `uid` to `taskUid`.
That's why I intentionally chose this example in my message hehe

And yes I think your suggestion is better than my initial message, thanks :+1:

Author: Ok, using gui's version then!

You may also query `GET` `/batches/:uid` to retrieve information about a single batch object:

```sh
curl  -X GET 'http://localhost:7700/batches/BATCH_UID'
```

`/batches/:uid` does not accept any parameters.

> 

Author: We stay simple in changelog, no need to add this 😊

Batch objects contain the following fields:

```json5
// TBD: an example of batch object
```

Additionally, task objects now include a new field, `batchUid`. Use this field together with `/batches/:uid` to retrieve data on a specific batch.

```json5
// TBD: an example of the new `task` object, one `progress` in implemented.

> ```

Author: @irevoire there is not `progress` yet. We need to add this example once it's done

> ```

Reviewer 4: There is no `progress` on the task.
Only the batches make progress.

This example was mostly to show the new `batchUid` field we added to the task object.

> ```

Author: Oh yes, my baaaaaad!

  "type": "documentAdditionOrUpdate",
  "canceledBy": null,
  "details": {
    "receivedDocuments": 1,
    "indexedDocuments": 1
  },
  "error": null,
  "duration": "PT0.027766819S",
  "enqueuedAt": "2024-12-02T14:07:34.974430765Z",
  "startedAt": "2024-12-02T14:07:34.99021667Z",
  "finishedAt": "2024-12-02T14:07:35.017983489Z"
}
```

Done by @irevoire in #5060, #5070, #5080

## Other improvements

* New query parameter for `GET` `/tasks`: `reverse`. If `reverse` is set to `true`, tasks will be returned in reversed order, from oldest to newest tasks. Done by @irevoire in #5048
* Phrase searches with`showMatchesPosition` set to `true` give a single location for the whole phrase @flevi29 in #4928
* New Prometheus metrics by @PedroTurik in #5044
* When a query finds matching terms in document fields with array values,  Meilisearch now includes an `indices` field to `_matchesPosition` specifying which array elements contain the matches by @LukasKalbertodt in #5005
* ⚠️ Breaking `vectorStore` change: field distribution no longer contains `_vectors`. Its value used to be incorrect, and there is no current use case for the fixed, most likely empty, value. Done as part of #4900

> 

Author: ```suggestion
* Improve error message by adding index name in #5056 by @airycanon

```

  * Fix the benchmarks by @irevoire in #5037
  * Bump Swatinem/rust-cache from 2.7.1 to 2.7.5 in #5030
  * Update charabia v0.9.2 by @ManyTheFish in #5098
  * Update mini-dashboard to v0.2.16 version by @curquiza in #5102
* CIs and tests
  * Improve performance of `delete_index.rs` by @DerTimonius in #4963
  * Improve performance of `create_index.rs` by @DerTimonius in #4962
  * Improve performance of `get_documents.rs` by @PedroTurik in #5025
  * Improve performance of `formatted.rs` by @PedroTurik in #5043
  * Fix the path used in the flaky tests CI by @Kerollmops in #5049
* Misc
  * Rollback the Meilisearch Kawaii logo by @Kerollmops in #5017
  * Add image source label to Dockerfile by @wuast94 in #4990
  * Hide code complexity into a subfolder by @Kerollmops in #5016
  * Internal tool: implement offline upgrade from v1.10 to v1.11 by @irevoire in #5034
  * Meilisearch is now able to retrieve Katakana words from a Hiragana query by @tats-u in https://github.com/meilisearch/charabia/pull/312
  * Improve error handling when writing into LMDB by @Kerollmops in https://github.com/meilisearch/meilisearch/pull/5089

❤️ Thanks again to our external contributors:

> - [Meilisearch](https://github.com/meilisearch/meilisearch): @DerTimonius, @flevi29, @LukasKalbertodt, @PedroTurik, @wuast94

Author: ```suggestion
- [Meilisearch](https://github.com/meilisearch/meilisearch): @airycanon, @DerTimonius, @flevi29, @LukasKalbertodt, @PedroTurik, @wuast94
```

- `"indexingTime"`: enables prefix processing during indexing. This is the default Meilisearch behavior
- `"disabled"`: deactivates prefix search completely

Disabling prefix search means the query `he` will no longer match the word `hello`. This may significantly impact search result relevancy, but speeds up the indexing process.

## New API route: `/batches`

The new `/batches` endpoint allow you to query information about task batches.

`GET` `/batches` returns a list of batch objects:

```sh
curl  -X GET 'http://localhost:7700/batches'
```

This endpoint accepts the same parameters as `GET` `/tasks` route, allowing you to narrow down which batches you want to see. Parameters used with `GET` `/batches` apply to the tasks, not the batches themselves. For example, `GET /batches?uid=0` returns batches containing tasks with a `taskUid` of `0` , not batches with a `batchUid` of `0`.

You may also query `GET` `/batches/:uid` to retrieve information about a single batch object:

```sh
curl  -X GET 'http://localhost:7700/batches/BATCH_UID'
```

`/batches/:uid` does not accept any parameters.

Batch objects contain the following fields:

> ```json5
> // TBD: an example of batch object
> ```

Reviewer 4: ````suggestion
```json5
{
  "uid": 160,
  "progress": {
    "steps": [
      {
        "currentStep": "processing tasks",
        "finished": 0,
        "total": 2
      },
      {
        "currentStep": "indexing",
        "finished": 2,
        "total": 3
      },
      {
        "currentStep": "extracting words",
        "finished": 3,
        "total": 13
      },
      {
        "currentStep": "document",
        "finished": 12300,
        "total": 19546
      }
    ],
    "percentage": 37.986263
  },
  "details": {
    "receivedDocuments": 19547,
    "indexedDocuments": null
  },
  "stats": {
    "totalNbTasks": 1,
    "status": {
      "processing": 1
    },
    "types": {
      "documentAdditionOrUpdate": 1
    },
    "indexUids": {
      "mieli": 1
    }
  },
  "duration": null,
  "startedAt": "2024-12-12T09:44:34.124726733Z",
  "finishedAt": null
}
```
````

Meilisearch v1.12 introduces significant indexing speed improvements, almost halving the time required to index large datasets. This release also introduces new settings to customize and potentially further increase indexing speed.

<!-- The following lines should NOT be put in the PRE-release changelogs -->
🧰 All official Meilisearch integrations (including SDKs, clients, and other tools) are compatible with this Meilisearch release. Integration deployment happens between 4 to 48 hours after a new version becomes available.

<!-- The following lines should NOT be put in the PRE-release changelogs -->
Some SDKs might not include all new features. Consult the project repository for detailed information. Is a feature you need missing from your chosen SDK? Create an issue letting us know you need it, or, for open-source karma points, open a PR implementing it (we'll love you for that ❤️).

# New features and updates 🔥

## Improve indexing speed

> Indexing time for huge dataset import (multiple millions of documents) is divided by two!
> 
> TBD: give more data and metrics about improvements

Reviewer 3: ```suggestion
Indexing time is improved across the board!

- Performance is maintained or better on smaller machines
- On bigger machines with multiple cores and good IO, Meilisearch v1.12 is much faster than Meilisearch v1.11
    - More than twice as fast for raw document insertion tasks.
    - More than x4 as fast for incrementally updating documents in a large database.
    - Embeddings generation was also improved up to x1.5 for some workloads.
    
The new indexer also makes task cancellation faster.
```

Done by @dureuill, @ManyTheFish, and @Kerollmops in #4900.

## TBD: Customize settings to improve indexing speed

TBD

## More visibility around indexing tasks

In order to give more visibility around indexing processing around your `task`, especially when indexing big documents, new routes have been introduced:
- `GET /batches/:uid`: returns a specific `batch` object
- `GET /batches`: return a list of `batch` objects

The same query parameters than `GET /tasks` route can be used to apply filtering.

TBD: an example of batch object

Also, a new field is introduced in the task object: `batch`.

TBD: an example of the new `task` object, one `progress` in implemented.

Done by @irevoire in #5060, #5070, #5080

## Other improvements

* Introduce the `reverse` query parameter for `GET /tasks` route set to `false` by default. If set to `true`, then the tasks will be returned in reversed order (the oldest first). Done by @irevoire in #5048.
* Make matches consider phrases as a single `Match` by @flevi29 in #4928
* Adds new metrics to prometheus by @PedroTurik in #5044
* Add `indices` field to `_matchesPosition` to specify where in an array a match comes from, by @LukasKalbertodt in #5005

> 

Reviewer 3: ```suggestion
* Breaking for users of the experimental `vectorStore` feature: The field distribution no longer contains `_vectors`. Its value used to be incorrect, and fixing it gives it a very uninteresting value, as all documents in an index with an embedder have the `_vectors` field. Done as part of #4900

```

<!-- The following lines should NOT be put in the PRE-release changelogs -->
🧰 All official Meilisearch integrations (including SDKs, clients, and other tools) are compatible with this Meilisearch release. Integration deployment happens between 4 to 48 hours after a new version becomes available.

<!-- The following lines should NOT be put in the PRE-release changelogs -->
Some SDKs might not include all new features. Consult the project repository for detailed information. Is a feature you need missing from your chosen SDK? Create an issue letting us know you need it, or, for open-source karma points, open a PR implementing it (we'll love you for that ❤️).

# New features and updates 🔥

## Improve indexing speed

Indexing time for huge dataset import (multiple millions of documents) is divided by two!

TBD: give more data and metrics about improvements

Done by @dureuill, @ManyTheFish, and @Kerollmops in #4900.

## TBD: Customize settings to improve indexing speed

> TBD

Reviewer 5: ```suggestion
Two new opt-out settings, working as the `proximityPrecision` one, have been added `facetSearch` and `prefixSearch`.
The purpose of these new settings is to let the user deactivate some features to save indexing time.

### `facetSearch` setting
This setting activates or deactivates the [facet search feature](https://www.meilisearch.com/docs/learn/filtering_and_sorting/search_with_facet_filters#searching-facet-values) for all the filterable attributes of the index:
#### PUT `/indexes/{index_uid}/settings/facet-search`
Value: `true` or `false`

Changing this setting activates or deactivates the facet search feature

#### DELETE `/indexes/{index_uid}/settings/facet-search`
Reset the `facetSearch` setting to its default value: `true`

### `prefixSearch` setting
This setting activates or deactivates the ability to search a word by prefix on an index:
#### PUT `/indexes/{index_uid}/settings/prefix-search`
Value: `"disabled"` or `"indexingTime"`

- `"disabled"`: deactivate the prefix search completely
> During a search, the word `hello` will not be retrieved anymore with the query `he`, the prefixes will not be processed during the indexing, saving time. 

- `"indexingTime"`: Activate the processing of the prefixes (default Meilisearch behavior)

#### DELETE `/indexes/{index_uid}/settings/prefix-search`
Reset the `prefixSearch` setting to its default value: `"indexingTime"`

```

## More visibility around indexing tasks

In order to give more visibility around indexing processing around your `task`, especially when indexing big documents, new routes have been introduced:
- `GET /batches/:uid`: returns a specific `batch` object
- `GET /batches`: return a list of `batch` objects

The same query parameters than `GET /tasks` route can be used to apply filtering.

TBD: an example of batch object

Also, a new field is introduced in the task object: `batch`.

TBD: an example of the new `task` object, one `progress` in implemented.

Done by @irevoire in #5060, #5070, #5080

## Other improvements

* Introduce the `reverse` query parameter for `GET /tasks` route set to `false` by default. If set to `true`, then the tasks will be returned in reversed order (the oldest first). Done by @irevoire in #5048.
* Make matches consider phrases as a single `Match` by @flevi29 in #4928
* Adds new metrics to prometheus by @PedroTurik in #5044
* Add `indices` field to `_matchesPosition` to specify where in an array a match comes from, by @LukasKalbertodt in #5005

> * Breaking for users of the experimental `vectorStore` feature: The field distribution no longer contains `_vectors`. Its value used to be incorrect, and fixing it gives it a very uninteresting value, as all documents in an index with an embedder have a (potentially empty) `_vectors` field. Done as part of #4900

Author: ```suggestion
* ⚠️ Breaking for users of the experimental `vectorStore` feature: The field distribution no longer contains `_vectors`. Its value used to be incorrect, and fixing it gives it a very uninteresting value, as all documents in an index with an embedder have a (potentially empty) `_vectors` field. Done as part of #4900
```

# Fixes 🐞

* Return an appropriate error when primary keys is greater than 512 bytes, by @flevi29 in #4930

> 

Reviewer 5: ```suggestion
* Numbers are now segmented the same way by the tokenizer by @dqkqd in https://github.com/meilisearch/charabia/issues/271
* Fix pagination when embedding fails by @dureuill in https://github.com/meilisearch/meilisearch/pull/5063
* Fix The stop words are ignored by Meili
#4984 by @ManyTheFish 
* FIx Phrase search doesn't work with attributesToSearchOn
#4974 by @ManyTheFish 

```
